### PR TITLE
Move rustc docs to the book

### LINKED ISSUE #21281: librustc_borrowck/borrowck/doc.rs doesn't get rendered
Because `mod borrowck` is private.

### PR DESCRIPTION
**Pull request by steveklabnik**
This is super black magic internals at the moment, but having it
somewhere semi-public seems good. The current versions weren't being
rendered, and they'll be useful for some people.

Fixes #21281

r? @nikomatsakis @kmcallister

### CODE REVIEW

**Comment by huonw** [comment]
Hm, general note, the trpl directory is getting quite busy: https://github.com/steveklabnik/rust/tree/gh21281/src/doc/trpl

Is it possible to e.g. put these into an `internals` subdirectory?

**Comment by steveklabnik** [comment]
Maybe. It ends up making more nested URLs, and I don't remember if Rustbook supports that or not.

File: src/doc/trpl/inference.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+course, it depends on the program.
+
+The main case which fails today that I would like to support is:
+
+```text
+fn foo<T>(x: T, y: T) { ... }
+
+fn bar() {
+    let x: @mut int = @mut 3;
+    let y: @int = @3;
```
Comment (kmcallister) [inline src/doc/trpl/inference.md]: Wow, this is a blast from the past.

File: src/doc/trpl/region-inference.md [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+contracting nodes. However, as I said, this does not work given the
+current treatment of closure bounds, but perhaps in the future we can
+address this problem somehow and make region inference somewhat more
+efficient. Note that this is solely a matter of performance, not
+expressiveness.
+
+### Skolemization
+
+For a discussion on skolemization and higher-ranked subtyping, please
+see the module `middle::infer::higher_ranked::doc`.
```
Comment (kmcallister) [inline src/doc/trpl/region-inference.md]: Should become a HTML link.

**Comment by kmcallister** [comment]
Looks good to me.  I wholeheartedly approve of putting these in the book.  It's great to have a peek behind the curtain even if the only takeaway is "o_O".

In particular I think the borrowchecker docs are important if you want a precise understanding of Rust semantics.  It's more conceptual than a purely internal thing.  It might even be a good starting point for an "advanced borrowchecker" chapter in Part III.

(Actually, what I really would like to do is implement a toy borrowchecker for a subset of Rust, as a way to solidify my own understanding, and produce a blog series / literate code repo / something along the way.  But who knows when I'll have time to do that...)

btw, I agree with the choice to keep `trans` doc in the source tree.  That stuff is internal details even compared to the rest of these docs.

**Comment by kmcallister** [comment]
[This line](https://github.com/rust-lang/rust/pull/22126/files#diff-02ff348c7221ee798f0c755c3aa2bdddR299) should also become a link.

**Comment by kmcallister** [comment]
Might want to add an informative sidebar somewhere on the subject of "What the hell does 'skolemize' mean?"

File: src/doc/trpl/inference.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/trpl/inference.md]: heh, yeah, I was solely looking to move stuff, not to update it.

**Comment by nikomatsakis** [comment]
Hmm, I don't know how I feel about having these outside the compiler source code. The whole idea is that when browsing code in directory `foo` you can look for a `doc.rs` to get a more detailed look. That is, they are extended code documents.

**Comment by nikomatsakis** [comment]
(They are already somewhat dated and ever at risk for being out of date, putting them in "the book" makes that even more likely it seems to me.)

**Comment by kmcallister** [comment]
> The whole idea is that when browsing code in directory foo you can look for a doc.rs to get a more detailed look.

In that case maybe they should be named `README.md` so you can browse them on GitHub.  Rendering through rustdoc didn't work that well, I guess because some of these modules are private.

**Comment by steveklabnik** [comment]
Good idea, @kmcallister . Done. What do you think, @nikomatsakis ?

**Comment by kmcallister** [comment]
Should also help them stand out in the directory listing.

**Comment by nikomatsakis** [comment]
seems good to me!

**Comment by steveklabnik** [comment]
Oops, forgot to remove a few `mod doc` lines.

**Pull request merged**